Src/Particle/AMReX_WriteBinaryParticleData.H

```diff
@@ -24,6 +27,284 @@ std::size_t PSizeInFile (const Vector<int>& wrc, const Vector<int>& wic)
+countFlags (const Vector<std::map<std::pair<int,int>,Container<int,Allocator>>>& particle_io_flags,
+            const PC& pc, amrex::Long& nparticles)
+{
+    for (int lev = 0; lev < pc.GetParticles().size();  lev++)
+    {
+        const auto& pmap = pc.GetParticles(lev);
+        for (const auto& kv : pmap)
+        {
+            const auto& pflags = particle_io_flags[lev].at(kv.first);
+            for (int k = 0; k < kv.second.numParticles(); ++k)
+            {
+                if (pflags[k]) nparticles++;
+            }
+        }
+    }
+}
+
+template <template <class, class> class Container, class Allocator>
+typename std::enable_if<!RunOnGpu<typename Container<int, Allocator>::allocator_type>::value>::type
+countFlags (const Container<int,Allocator>& pflags, int& nparticles)
+{
+    for (std::size_t k = 0; k < pflags.size(); ++k)
+    {
+        if (pflags[k]) nparticles++;
+    }
+}
+
+template <class PC>
+typename std::enable_if<RunOnGpu<typename PC::template AllocatorType<int>>::value>::type
+packIOData (Vector<int>& idata, Vector<ParticleReal>& rdata, const PC& pc, int lev, int grid,
+            const Vector<int>& write_real_comp, const Vector<int>& write_int_comp,
+            const Vector<std::map<std::pair<int, int>, typename PC::IntVector>>& particle_io_flags,
+            const Vector<int>& tiles, int np)
+{
+    int num_output_int = 0;
+    for (int i = 0; i < pc.NumIntComps() + PC::NStructInt; ++i)
+        if (write_int_comp[i]) ++num_output_int;
+
+    const int iChunkSize = 2 + num_output_int;
+    idata.resize(np*iChunkSize);
+
+    int num_output_real = 0;
+    for (int i = 0; i < pc.NumRealComps() + PC::NStructReal; ++i)
+        if (write_real_comp[i]) ++num_output_real;
+
+    const int rChunkSize = AMREX_SPACEDIM + num_output_real;
+    rdata.resize(np*rChunkSize);
+
+    typename PC::IntVector  idata_d(idata.size());
+    typename PC::RealVector rdata_d(rdata.size());
+
+    typename PC::IntVector write_int_comp_d(write_int_comp.size());
+    typename PC::IntVector write_real_comp_d(write_real_comp.size());
+    Gpu::copy(Gpu::hostToDevice, write_int_comp.begin(), write_int_comp.end(),
+              write_int_comp_d.begin());
+    Gpu::copy(Gpu::hostToDevice, write_real_comp.begin(), write_real_comp.end(),
+              write_real_comp_d.begin());
+    Gpu::Device::synchronize();
+
+    const auto write_int_comp_d_ptr = write_int_comp_d.data();
+    const auto write_real_comp_d_ptr = write_real_comp_d.data();
+
+    std::size_t poffset = 0;
+    for (int i = 0; i < tiles.size(); i++) {
+        const auto& ptile = pc.ParticlesAt(lev, grid, tiles[i]);
+        const auto& pflags = particle_io_flags[lev].at(std::make_pair(grid, tiles[i]));
+        int np_tile = ptile.GetArrayOfStructs().numParticles();
+        typename PC::IntVector offsets(np_tile);
+        Gpu::exclusive_scan(pflags.begin(), pflags.end(), offsets.begin());
```

Comment:
Potentially we could save one memcpy by using ExclusiveSum that will return the total sum.